src/passes/ifStatementTempVarPostpender.ts

```diff
@@ -0,0 +1,194 @@
+ * }
+ * g(y) // reference to y
+ ********************
+ *
+ * Becomes:
+ *
+ ********************
+ * // y is not live
+ * let y = 0
+ * // y is live
+ * if (x == 0) {
+ *   // y is live
+ *   tempvar y = y;
+ * } else {
+ *   // y is not live
+ *   let y = 4;
+ *   // y is live
+ *   tempvar y = y;
+ * }
+ * tempvar y = y;
+ * g(y) // reference to y
+ ********************
+ */
+
+export class IfStatementTempVarPostpender extends ASTMapper {
+  private liveVars: Set<string> = new Set();
+
+  // Function to add passes that should have been run before this pass
+  addInitialPassPrerequisites(): void {
+    const passKeys: Set<string> = new Set<string>([
+      'An', // AnotateImplicit pass creates cairoFunctionDefinitions which this pass use
+    ]);
+    passKeys.forEach((key) => this.addPassPrerequisite(key));
+  }
+
+  visitCairoFunctionDefinition(node: CairoFunctionDefinition, ast: AST): void {
+    this.liveVars = new Set<string>();
+    [...node.implicits].forEach((implicit) => this.liveVars.add(implicit));
+    // analyse the block
+    if (node.vBody) {
+      this.dispatchVisit(node.vBody, ast);
+    }
+    // remove the parameters and implicits
+    [...node.implicits].forEach((implicit) => {
+      if (!this.liveVars.delete(implicit)) {
+        throw new TranspileFailedError(
+          `Implicit variable dead. ${implicit} in function ${node.name}.`,
+        );
+      }
+    });
+    this.liveVars = new Set<string>();
+  }
+
+  visitIdentifier(node: Identifier, _ast: AST): void {
+    if (
+      node.vReferencedDeclaration !== undefined &&
+      node.vReferencedDeclaration.getClosestParentByType(CairoFunctionDefinition) !== undefined
+    ) {
+      this.liveVars.add(node.name);
+    }
+  }
+
+  visitBlock(node: Block, ast: AST): void {
+    node.vStatements
+      .slice(0)
+      .reverse()
+      .map((child) => this.dispatchVisit(child, ast));
+  }
+
+  visitUncheckedBlock(node: UncheckedBlock, ast: AST): void {
+    node.vStatements
+      .slice(0)
+      .reverse()
+      .map((child) => this.dispatchVisit(child, ast));
+  }
+
+  visitVariableDeclarationStatement(node: VariableDeclarationStatement, ast: AST): void {
+    node.vDeclarations.forEach((declaration) => this.liveVars.delete(declaration.name));
+    if (node.vInitialValue) {
+      this.dispatchVisit(node.vInitialValue, ast);
+    }
+  }
+
+  visitFunctionCall(node: FunctionCall, ast: AST): void {
+    // ignore the function identifier
+    for (const arg of node.vArguments) {
+      this.dispatchVisit(arg, ast);
+    }
+  }
+
+  visitIfStatement(node: IfStatement, ast: AST): void {
+    ensureBothBranchesAreBlocks(node, ast);
+    assert(node.vFalseBody instanceof Block);
+
+    const liveVarsBefore = new Set(this.liveVars);
+    this.dispatchVisit(node.vTrueBody, ast);
+    const liveVarsAfterTrueBody = new Set(this.liveVars);
+
+    this.liveVars = new Set(liveVarsBefore);
+    let liveVarsAfterFalseBody = new Set<string>();
+    if (node.vFalseBody) {
```

Comment:
With the previous assertion this condition should always be true, so I don't think it is necessary